### PR TITLE
optimized some grid reorder function

### PR DESCRIPTION
**Pull request by hfp**
This mainly improves code generation with GCC.

### CODE REVIEW

**Comment by hfp** [comment]
Merging this PR shall wait until we have Dashboard MC results of previous changes.

**Comment by alazzaro** [comment]
> Merging this PR shall wait until we have Dashboard MC results of previous changes.

Do you mean the MC tests on Daint?

**Comment by hfp** [comment]
> Do you mean the MC tests on Daint?

Yes, unless we want to judge my performance claim differently. It should at least show no regression on top of Mathieu's changes. Although, now that I write this comment - ref-backend shall not be affected by M.'s change...

**Comment by mtaillefumier** [comment]
Short question. The dashboard uses the regtests inputs files or not ? If
so none of my code should be called at all since it always use the
reference backend. Of course compilation issues are a separate issue.

**Comment by alazzaro** [comment]
Just started the test ;)

**Comment by mtaillefumier** [comment]
On 11/20/20 11:29 AM, Hans Pabst wrote:
>
>     Do you mean the MC tests on Daint?
>
> Yes, unless we want to judge my performance claim differently. It
> should at least show no regression on top of Mathieu's changes.
> Although, now that I write this comment - ref-backend shall not be
> affected by M.'s change...
>
No they should not. The only thing I modified is the use of enum instead
of #define. And the two changes of the new code should not either.

**Comment by alazzaro** [comment]
The daint tests are using some of the tests from:

https://github.com/cp2k/cp2k/tree/master/benchmarks

So, we are not changing the input...

**Comment by hfp** [comment]
> Just started the test ;)

> No they should not.

Ok, I merge this PR when green since Alfio triggered the latest MC perf test. With this PR merged, we should at least see no degradation.

**Comment by mtaillefumier** [comment]
Thanks. Then they should pass. If they don't it means something else is
going on.

**Comment by mtaillefumier** [comment]
If the code compiles. If libxsmm is not used then the compilation will
fail because of the unused parameter warning treated as error. I have
the workaround already. inlined functions as well but it was fixed already.

File: src/grid/ref/grid_ref_collint.h [hunk 1]
```diff
@@ -396,13 +396,13 @@ static inline void general_ci_to_reg(const int lp, const double exp_aiibic,
...
-                   GRID_CONST_WHEN_INTEGRATE double *grid) {
+static inline void general_ci_to_grid(
+    const int lp, const int jg, const int kg, const int ismin, const int ismax,
+    const int npts_local[3], const int bounds_i[2], const int index_min[3],
+    const int map_i[], const double zetp, const double a, const double b,
+    const double c, const double exp_ab, const double exp_2a,
+    const double gp[3], GRID_CONST_WHEN_COLLOCATE double *ci,
+    GRID_CONST_WHEN_INTEGRATE double *grid) {
 
   //--------------------------------------------------------------------
```
Comment (oschuett) [inline src/grid/ref/grid_ref_collint.h]: Maybe this comment should be moved alongside.

**Comment by alazzaro** [comment]
Test completed: https://dashboard.cp2k.org/archive/cray-xc50-daint-mc-psmp_perf/index.html
Seems everything OK

File: src/grid/ref/grid_ref_collint.h (hunk 1 above)
Comment (hfp) [reply src/grid/ref/grid_ref_collint.h]: Yes, that was my concern too. I will put attach it to the call-side where this work now takes place.

**Pull request merged**